File: leetcode/src/java/ValidPalindrome.java
public class ValidPalindrome {
    public static void main(String[] args) {
        ValidPalindrome instance = new ValidPalindrome();
        System.out.println(instance.isPalindrome("A man, a plan, a canal: Panama"));
        System.out.println(instance.isPalindrome("race a car"));
        System.out.println(instance.isPalindrome(" "));
        System.out.println(instance.isPalindrome(",:"));
        System.out.println(instance.isPalindrome("0P"));
    }
    public boolean isPalindrome(String s) {
        boolean result = true;
        int start = 0;
        int end = s.length() - 1;
        while (start < end) {
            while (start <= end && !(isChar(s.charAt(start)) || isNum(s.charAt(start)))) {
                start++;
            }
            while (start <= end && !(isChar(s.charAt(end)) || isNum(s.charAt(end)))) {
                end--;
            }
            if (start > end) {
                break;
            } else {
                if (s.charAt(start) == s.charAt(end) || (isChar(s.charAt(start)) && isChar(s.charAt(end)) && Math.abs(s.charAt(start) - s.charAt(end)) == 32)) {
                    start++;
                    end--;
                } else {
                    result = false;
                    break;
                }
            }
        }
        return result;
    }
    private boolean isChar(char c) {
        return ((65 <= c && c <= 90) || (97 <= c && c <= 122));
    }
    private boolean isNum(char c) {
        return 48 <= c && c <= 57;
    }
}

File: leetcode/src/python/valid_palindrome.py
class ValidPalindrome:
    def isPalindrome(self, s: str) -> bool:
        new_s = ''
        for i, c in enumerate(s):
            if c.isalnum():
                new_s += c.lower()

        for i in range(len(new_s)):
            if i <= len(new_s) - 1 - i and new_s[i] != new_s[len(new_s) - 1 - i]:
                return False
        return True

if __name__ == '__main__':
    instance = ValidPalindrome()
    #s = 'A man, a plan, a canal: Panama'
    s = '0P'
    print(instance.isPalindrome(s))